htdoc/htfuzzy.1.in
.TH htfuzzy 1 "21 July 1997"
.\" NAME should be all caps, SECTION should be 1-8, maybe w/ subsection
.\" other parms are allowed: see man(7), man(1)
.SH NAME
htfuzzy \- creates indexes for different ``fuzzy'' search algorithms for
the ht:://Dig search engine
.SH SYNOPSIS
.B htfuzzy
.I "[options] algorithm ..."
.SH "DESCRIPTION"
.I Htfuzzy
creates indexes for different ``fuzzy'' search algorithms.
These indexes can then be used by the
.I htsearch
program.
.SH OPTIONS
.TP
.B \-c \fIconfigfile\fR
Use the specified
.I configfile
instead of the default. 
.TP
.B \-v
Verbose mode.
Used once will provide progress feedback, used
more than once will overflow even the biggest buffers. :-) 
.SH "ALGORITHMS"
Indexes for the following search algorithms can currently be created: 
.TP
.B soundex 
Creates a slightly modified soundex key database. Differences
with the standard soundex algorithm are: Keys are 6 digits and the first
letter is also encoded. 
.TP
.B metaphone 
Creates a metaphone key database. This algorithm is more
specific to English, but will get fewer "weird" matches than the
soundex algorithm. 
.TP
.B endings 
Creates two databases which can be used to match common
word endings. The creation of these databases requires a list of
affix rules and a dictionary which uses those affix rules. The
format of the affix rules and dictionary files are the ones used
by the ispell program. Included with the distribution are the
affix rules for English and a fairly small English dictionary.
Other languages can be supported by getting the appropriate
affix rules and dictionaries. These are available for many
languages; check the ispell distribution for more details. 
.TP
.B synonyms 
Creates a database of synonyms for words. It reads a text
database of synonyms and creates a database that htsearch
can then use. Each line of the text database consists of words
where the first word will have the other words on that line as
synonyms. 
.SH "FILES"
.TP
.B @DEFAULT_CONFIG_FILE@
The default configuration file.
.SH "SEE ALSO"
Please refer to the HTML pages (in the htdig-doc package)
.B (THIS MUST BE CUSTOMISED...)
.B /usr/share/doc/htdig-doc/html/index.html
and the manual pages
.I htdig(1)
,
.I htmerge(1)
, and 
.I htsearch(1)
for a detailed description of ht://Dig and its commands.
.SH AUTHOR
This manual page was written by Christian Schwarz, modified by
Stijn de Bekker, based on the HTML documentation of ht://Dig.


ChangeLog
Wed Feb 23 16:53:40 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htcommon/defaults.cc (backlink_factor, external_parser,
	local_default_doc, local_urls, local_urls_only, local_user_urls):
	Add some updates from 3.1.5's attrs.html.
	* htdoc/attrs.html, htdoc/cf_byname.html, htdoc/cf_byprog.html:
	Regenerate using cf_generate.pl.

Wed Feb 23 15:11:51 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	[ Improve htsearch's HTML 4.0 compliance ]
	* htsearch/TemplateList.cc (createFromString): Use file name rather
	than internal name to select builtin-* templates, use $&(TITLE) and
	$&(URL) in templates and quote HTML tag parameters.
	* installdir/long.html, installdir/short.html: Use $&(TITLE) and
	$&(URL) in templates and quote HTML tag parameters.
	* htsearch/Display.cc (setVariables): quote all HTML tag parameters
	in generated select lists.
	* installdir/footer.html, installdir/header.html,
	installdir/nomatch.html, installdir/search.html,
	installdir/syntax.html, installdir/wrapper.html:
	Use $&(var) where appropriate, and quote HTML tag parameters.
	* installdir/htdig.conf: quote all HTML tag parameters.

Wed Feb 23 13:40:27 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htcommon/URL.h (encodeURL): Change list of valid characters to
	include only unreserved ones.
	* htcommon/cgi.cc (init): Allow "&" and ";" as input param. separators.
	* htsearch/Display.cc (createURL): Encode each parameter separately,
	using new unreserved list, before piecing together query string, to
	allow characters like "?=&" within parameters to be encoded.

Wed Feb 23 13:22:29 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htlib/URL.cc (ServerAlias): Fix server_aliases processing to prevent
	infinite loop (as for local_urls in PR#688).

Wed Feb 23 12:49:52 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htlib/HtDateTime.h, htlib/HtDateTime.cc: change Httimegm() method
	to HtTimeGM(), to avoid conflict with Httimegm() C function, so we
	don't need "::" override, for Mac OS X.
	* htlib/htString.h, htlib/String.cc: change write() method to
	Write(), to avoid conflict with write() function, so we don't need
	"::" override, for Mac OS X.

Wed Feb 23 12:17:46 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htlib/Configuration.cc(Read): Fixed to allow final line without
	terminating newline character, rather than ignoring it.

Wed Feb 23 12:01:01 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htdig/Retriever.cc (GetLocal, GetLocalUser): Add URL-decoding
	enhancements to local_urls, local_default_urls & local_default_doc,
	to allow hex encoding of special characters.

Wed Feb 23 19:14:29 2000  Loic Dachary  <loic@ceic.com>

 	* htcommon/conf_parser.cxx: regenerated from conf_parser.yxx

Wed Feb 23 19:04:16 2000  Loic Dachary  <loic@ceic.com>

	* test/test_functions.in: inconditionaly remove existing test/var
	  directory before runing tests to prevent accidents.

	* htcommon/URL.cc (URL): fixed String->char warning

	* htcommon/defaults.cc (wordlist_compress): defaults to true

Tue Feb 22 17:09:10 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htdig/HTML.cc(parse, do_tag): Fix handling of <img alt=...> text
	and parsing of words in meta tags, to to proper word separation.
	* htlib/HtWordType.h, htlib/HtWordType.cc: Add HtWordToken() function,
	to replace strtok() in HTML parser.

Tue Feb 22 16:21:25 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htlib/URL.cc (ctor, normalizePath): Fix PR#779, to handle relative
	URLs correctly when there's a trailing ".." or leading "//".

Tue Feb 22 14:09:26 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htdig/Document.cc (RetrieveLocal): Handle common extensions for
	text/plain, application/pdf & application/postscript.

Mon Feb 21 17:25:21 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* contrib/htdig-3.2.0.spec: Fixed %post script to add more
	descriptive entries in htdig.conf, made cron script a config file,
	updated to 3.2.0b2.

	* contrib/conv_doc.pl, contrib/parse_doc.pl: Added comments to show
	Warren Jones's updates in change history.

Mon Feb 21 17:09:13 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htcommon/HtConfiguration.h, htcommon/conf_parser.yxx,
	htlib/Configuration.h, htlib/Configuration.cc: split Add() method
	into Add() and AddParsed(), so that only config attributes get parsed.
	Use AddParsed() only in Read() and Defaults().

Sat Feb 19 04:56:47 2000  Geoff Hutchison  <ghutchis@wso.williams.edu>

	* htnet/Connection.h, htnet/Connection.cc: Renamed methods with
	capitals to remove the need to use ::-escaped library calls.

Fri Feb 18 14:40:50 2000  Loic Dachary  <loic@ceic.com>

	* test/conf/access.conf.in: removed cookies. Not used and some
	  httpd are not compiled with usertrack.

Wed Feb 16 12:15:08 2000  Vadim Chekan  <vadim@etc.lviv.ua>

	* htcommon/Makefile.am replaced conf.tab.cc.h by conf_parser.h in
	noinst_HEADERS

	* htcommon/conf_parser.yxx,conf_parser.lxx,HtConfiguration.cc,
	HtConfiguration.h: added copyright and Id:

	* htcommon/cgi.cc(init): fixed bug: array must be free by 
	delete [] buf, not just delete buf;

Tue Feb 15 23:16:14 2000  Geoff Hutchison  <ghutchis@wso.williams.edu>

	* htnet/HtHTTP.cc (isParsable): Remove application/pdf as a
	default type--it is now handled through the ExternalParser
	interface if at all.

	* htcommon/defaults.cc: Remove pdf_parser attribute.

	* htdoc/attrs.html, htdoc/cf_byname.html, htdoc/cf_byprog.html:
	Regenerate using cf_generate.pl.

	* htdig/Document.cc (getParsable): Remove PDF once and for all
	(hopefully).

	* htdig/ExternalParser.cc (parse): Ditto.

	* configure.in: Remove check for PDF_PARSER.

	* configure: Regenerate using autoconf

	* htdig/Makefile.am: Remove PDF.cc and PDF.h.

	* Makefile.in, */Makefile.in: Regenerate using automake --foreign

Tue Feb 15 12:02:39 EET 2000  Vadim Chekan  <vadim@etc.lviv.ua>

	* htcommon/HtConfiguration.cc,HtConfiguration.h: fixed bug discovered

htfuzzy/htfuzzy.cc
//
// htfuzzy.cc
//
// Create one or more ``fuzzy'' indexes into the main word database.
// These indexes can be used by htsearch to perform a search that uses
// other algorithms than exact word match.
//
// This program is meant to be run after htmerge has created the word
// database.
//
// For each fuzzy algorithm, there will be a separate database.  Each
// database is simply a mapping from the fuzzy key to a list of words
// in the main word database.
//
// $Log: htfuzzy.cc,v $
// Revision 1.9  1999/01/07 03:13:49  ghutchis
// Fix minor memory leaks.
//
// Revision 1.8  1998/12/04 04:13:51  ghutchis
// Use configure check to only include getopt.h when it exists.
//
// Revision 1.6  1998/10/12 02:04:00  ghutchis
//
// Updated Makefiles and configure variables.
//
// Revision 1.4  1998/09/18 02:38:08  ghutchis
//
// Bug fixes for 3.1.0b2
//
// Revision 1.3  1997/06/23 02:26:15  turtle
// Added version info the usage output
//
// Revision 1.2  1997/03/24 04:33:19  turtle
// Renamed the String.h file to htString.h to help compiling under win32
//
// Revision 1.1.1.1  1997/02/03 17:11:12  turtle
// Initial CVS
//
//
#if RELEASE
static char RCSid[] = "$Id: htfuzzy.cc,v 1.9 1999/01/07 03:13:49 ghutchis Exp $";
#endif

#include "htfuzzy.h"
#include "Fuzzy.h"
#include "Soundex.h"
#include "Endings.h"
#include "Metaphone.h"
#include "Synonym.h"
#include <htString.h>
#include <List.h>
#include <Dictionary.h>
#include <defaults.h>

// If we have this, we probably want it.
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif

int		debug = 0;

void usage();


//*****************************************************************************
// int main(int ac, char **av)
//
int
main(int ac, char **av)
{
    int			c, i;
    extern char	*optarg;
    extern int	optind;
    String		configFile = DEFAULT_CONFIG_FILE;

    //
    // Parse command line arguments
    //
    while ((c = getopt(ac, av, "c:v")) != -1)
    {
	switch (c)
	{
	    case 'c':
		configFile = optarg;
		break;
				
	    case 'v':
		debug++;
		break;
				
	    default:
		usage();
	}
    }

    //
    // Determine what algorithms to use
    //
    List	wordAlgorithms;
    List	noWordAlgorithms;
    for (i = optind; i < ac; i++)
    {
	if (mystrcasecmp(av[i], "soundex") == 0)
	{
	    wordAlgorithms.Add(new Soundex);
	}
	else if (mystrcasecmp(av[i], "metaphone") == 0)
	{
	    wordAlgorithms.Add(new Metaphone);
	}
	else if (mystrcasecmp(av[i], "endings") == 0)
	{
	    noWordAlgorithms.Add(new Endings);
	}
	else if (mystrcasecmp(av[i], "synonyms") == 0)
	{
	    noWordAlgorithms.Add(new Synonym);
	}
	else
	{
	    reportError(form("'%s' is not a supported algorithm",
			     av[i]));
	}
    }
    if (wordAlgorithms.Count() == 0 && noWordAlgorithms.Count() == 0)
    {
	cout << "htfuzzy: No algorithms specified\n";
	usage();
    }
	
    //
    // Find and parse the configuration file.
    //
    config.Defaults(&defaults[0]);
    if (access(configFile, R_OK) < 0)
    {
	reportError(form("Unable to find configuration file '%s'",
			 configFile.get()));
    }
    config.Read(configFile);

    Fuzzy	*fuzzy;
    if (wordAlgorithms.Count() > 0)
    {
        //
        // Open the word database so that we can grab the words from it.
        //
        String  wordFile = config["word_db"];
        if (access(wordFile, R_OK) < 0)
        {
            reportError(form("Unable to read word database file '%s'\nDid you run htmerge?",
                             wordFile.get()));
        }
	Database *worddb = Database::getDatabaseInstance();
	if (worddb->OpenRead(wordFile) == OK)
	{
	    //
	    // Go through all the words in the database
	    //
	    char		*key;
	    Fuzzy		*fuzzy = 0;
	    String		word, fuzzyKey;
	    int			count = 0;

	    worddb->Start_Get();
	    while ((key = worddb->Get_Next()))
	    {
		word = key;
		wordAlgorithms.Start_Get();
		while ((fuzzy = (Fuzzy *) wordAlgorithms.Get_Next()))
		{
		    fuzzy->addWord(word);
		}
		count++;
		if ((count % 100) == 0 && debug)
		{
		    cout << "htfuzzy: words: " << count << '\n';
		    cout.flush();
		}
	    }
	    if (debug)
	    {
		cout << "htfuzzy: total words: " << count << "\n";
		cout << "htfuzzy: Writing index files...\n";
	    }

	    //
	    // All the information is now in memory.
	    // Write all of it out to the individual databases
	    //
	    wordAlgorithms.Start_Get();
	    while ((fuzzy = (Fuzzy *) wordAlgorithms.Get_Next()))
	    {
		fuzzy->writeDB(config);
	    }
	    worddb->Close();
	    delete worddb;
	    if (fuzzy)
	    	delete fuzzy;
	}
	else
	{
	    reportError("Unable to open word database");
	}
    }
    if (noWordAlgorithms.Count() > 0)
    {
	noWordAlgorithms.Start_Get();
	while ((fuzzy = (Fuzzy *) noWordAlgorithms.Get_Next()))
	{
	    if (debug)
	    {
		cout << "htfuzzy: Selected algorithm: " << fuzzy->getName()
		     << endl;
	    }
	    fuzzy->createDB(config);
	}
    }
	
    if (debug)
    {
	cout << "htfuzzy: Done.\n";
    }
	
    return 0;
}


//*****************************************************************************
// void usage()
//
void
usage()
{
    cout << "usage: htfuzzy [-c configfile][-v] algorithm ...\n";
    cout << "This program is part of ht://Dig " << VERSION << "\n\n";
    cout << "Supported algorithms:\n";
    cout << "\tsoundex\n";
    cout << "\tmetaphone\n";
    cout << "\tendings\n";
    cout << "\tsynonyms\n";
    cout << "\n";
	
    cout << "Options:\n";

    cout << "\t-c configfile\n";
    cout << "\t\tUse the specified configuration file instead of the\n";
    cout << "\t\tdefault.\n\n";

    cout << "\t-v\tVerbose mode.  This increases the verbosity of the\n";
    cout << "\t\tprogram.  Using more than 2 is probably only useful\n";
    cout << "\t\tfor debugging purposes.\n\n";

    exit(0);
}


//*****************************************************************************
// void reportError(char *msg)
//
void
reportError(char *msg)
{
    cout << "htfuzzy: " << msg << "\n\n";
    exit(1);
}


